pkg/querier/querier.go

```diff
@@ -153,19 +158,25 @@ func NewQueryable(dq, cq storage.Queryable, distributor Distributor, cfg Config)
 }
 
 type querier struct {
+	// primary querier is used for labels and metadata queries
+	primary  storage.Querier
 	queriers []storage.Querier
 
-	distributor Distributor
+	chunkIterFn chunkIteratorFunc
 	ctx         context.Context
 	mint, maxt  int64
 }
 
 // Select implements storage.Querier.
 func (q querier) Select(sp *storage.SelectParams, matchers ...*labels.Matcher) (storage.SeriesSet, storage.Warnings, error) {
 	// Kludge: Prometheus passes nil SelectParams if it is doing a 'series' operation,
-	// which needs only metadata.
+	// which needs only metadata. Here we expect that primary querier will handle that.
```

Comment:
> Here we expect that primary querier will handle that.

I would explain that, given the metadata queries have no mint/maxt and given it's not feasible to query the entire history, in Cortex we pick the trade-off to query only ingesters, so we skip querying the long-term storage.

Reply:
Good point. I wasn't 100% sure *why* we do it this way, only made sure to keep the functionality. I'll add your comment to the code.

Would it be possible to query long-time storage when using blocks? Queriers have that information available, don't they?

Reply:
> Would it be possible to query long-time storage when using blocks? Queriers have that information available, don't they?

It could be possible and Thanos does, but when I talked to Bartek about it he mentioned that Thanos is suffering the same issue as well and they may also want to introduce a limitation (Cortex limitation is quite strict because we query only active series, but still is a limitation). For now, I would stay consistent with the chunks storage.